File: algoexpert/easy/05.node_depth.js
#!/usr/bin/env node

class Node {
  constructor(value, left=null, right=null) {
    this.value = value;
    this.left = left;
    this.right = right;
    this.depth = 0;
  }
}

const bt =
      new Node(1,
               new Node(22,
                        new Node(4,
                                 new Node(8),
                                 new Node(9)),
                        new Node(5)),
               new Node(3,
                        new Node(6),
                        new Node(7)));

function calculate_depth_helper(bt, depth) {
  if (!bt) return depth;
  let current = 0;
  if(bt.left) current += calculate_depth_helper(bt.left, depth + 1);
  if(bt.right) current += calculate_depth_helper(bt.right, depth + 1);
  return depth + current;
}
function calculate_depth_recursive(bt) {
  return calculate_depth_helper(bt, 0);
}

function calculate_depth_iterative(bt) {
  let sum = 0;
  let depth = 0;
  let stack = [];
  stack.push(bt);
  while (stack.length) {
    let node = stack.pop();
    sum += node.depth;
    if (node.left) {
      node.left.depth = depth;
      stack.push(node.left);
    }
    if (node.right) {
      node.right.depth = depth;
      stack.push(node.right);
    }
    depth++;
  }
  return sum;
}

console.log(calculate_depth_recursive(bt));
console.log(calculate_depth_iterative(bt));

File: algoexpert/easy/05.node_depth.ml
type 'a tree =
  | Leaf
  | Node of 'a * 'a tree * 'a tree

let depth_sum tree =
  let rec f depth = function
    | Leaf -> 0
    | Node (_,left,right) ->
       depth + f (depth+1) right + f (depth+1) left
  in
  f 0 tree

let atree =
  Node(1,
       Node(2,
            Node(4,
                 Node(8,Leaf,Leaf),
                 Node(9,Leaf,Leaf)),
            Node(5,Leaf,Leaf)),
       Node(3,
            Node(6,Leaf,Leaf),
            Node(7,Leaf,Leaf))
    )
